File: dawn/src/main/java/com/lzq/dawn/util/thread/LinkedBlockingQueue4Util.java
package com.lzq.dawn.util.thread;

import androidx.annotation.NonNull;

import java.util.concurrent.LinkedBlockingQueue;

/**
 * @Name :LinkedBlockingQueue4Util
 * @Time :2022/9/2 14:48
 * @Author :  Lzq
 * @Desc :
 */
public class LinkedBlockingQueue4Util extends LinkedBlockingQueue<Runnable> {

    private volatile ThreadPoolExecutor4Util mPool;

    private int mCapacity = Integer.MAX_VALUE;

    LinkedBlockingQueue4Util() {
        super();
    }

    LinkedBlockingQueue4Util(boolean isAddSubThreadFirstThenAddQueue) {
        super();
        if (isAddSubThreadFirstThenAddQueue) {
            mCapacity = 0;
        }
    }

    LinkedBlockingQueue4Util(int capacity) {
        super();
        mCapacity = capacity;
    }

    @Override
    public boolean offer(@NonNull Runnable runnable) {
        if (mCapacity <= size() &&
                mPool != null && mPool.getPoolSize() < mPool.getMaximumPoolSize()) {
            // create a non-core thread
            return false;
        }
        return super.offer(runnable);
    }

    public void setmPool(ThreadPoolExecutor4Util mPool) {
        this.mPool = mPool;
    }

    public ThreadPoolExecutor4Util getmPool() {
        return mPool;
    }
}

File: dawn/src/main/java/com/lzq/dawn/util/thread/LinkedBlockingQueue4Util.kt
package com.lzq.dawn.util.thread

import java.util.concurrent.LinkedBlockingQueue

/**
 * @Name :LinkedBlockingQueue4Util
 * @Time :2022/9/2 14:48
 * @Author :  Lzq
 * @Desc :
 */
class LinkedBlockingQueue4Util : LinkedBlockingQueue<Runnable> {
    @Volatile
    private var mPool: ThreadPoolExecutor4Util? = null
    private var mCapacity = Int.MAX_VALUE

    internal constructor() : super()
    internal constructor(isAddSubThreadFirstThenAddQueue: Boolean) : super() {
        if (isAddSubThreadFirstThenAddQueue) {
            mCapacity = 0
        }
    }

    internal constructor(capacity: Int) : super() {
        mCapacity = capacity
    }

    override fun offer(runnable: Runnable): Boolean {
        return if (mCapacity <= size && mPool != null && mPool!!.poolSize < mPool!!.maximumPoolSize) {
            // create a non-core thread
            false
        } else super.offer(
            runnable
        )
    }

    fun setmPool(mPool: ThreadPoolExecutor4Util?) {
        this.mPool = mPool
    }

    fun getmPool(): ThreadPoolExecutor4Util? {
        return mPool
    }
}